src/users/users.service.ts

```diff
@@ -0,0 +1,188 @@
+// import { RedisService } from '../redis/redis.service';
+import { CACHE_MANAGER } from '@nestjs/cache-manager';
+
+@Injectable()
+export class UsersService {
+  constructor(
+    @Inject(CACHE_MANAGER)
+    private cacheManager: Cache,
+    @InjectRepository(User)
+    private userRepository: Repository<User>,
+    @InjectRepository(Login)
+    private loginRepository: Repository<Login>,
+    @InjectRepository(Tag)
+    private tagRepository: Repository<Tag>,
+
+    // private readonly redisService: RedisService,
+    private readonly tokenService: AuthService,
+    private readonly s3Service: S3Service,
+  ) {}
+
+  async createUser(createUserDto: CreateUserDto): Promise<LoginSuccessDto> {
+    const { name, email, password } = createUserDto;
+    const isExisted = await this.userRepository.findOne({ where: { email } });
+    if (isExisted)
+      throw new HttpException('email already exists', HttpStatus.CONFLICT);
+    /* login and userDetail are restore to different table, 
+      so to create user need to create User obj , after user.id created, and Login token will be created.*/
+    const user = new User();
+    user.name = name;
+    user.email = email;
+    const saveUser = await this.userRepository.save(user);
+    const userId = saveUser.user_id;
+    const userLogin = new Login();
+    userLogin.user_id = userId;
+    userLogin.name = name;
+    userLogin.token = await bcrypt.hash(password, 10);
+    await this.loginRepository.save(userLogin);
+    const jwtToken = this.tokenService.generateToken(user.user_id);
+    const expiredTime = this.tokenService.EXPIRE_TIME;
+
+    return { user: saveUser, access_token: jwtToken, expiredTime };
+  }
+
+  async login(loginUserDto: LoginUserDto): Promise<LoginSuccessDto> {
+    const { email, password } = loginUserDto;
+    const user = await this.userRepository.findOne({ where: { email } });
+    if (!user) throw new UnauthorizedException();
+    const login = await this.loginRepository.findOne({
+      where: { user_id: user.user_id },
+    });
+    const isValid = await login.comparePassword(password);
+    if (!isValid) throw new UnauthorizedException();
+
+    const jwtToken = this.tokenService.generateToken(user.user_id);
+    const expiredTime = this.tokenService.EXPIRE_TIME;
+    return { user, access_token: jwtToken, expiredTime };
+  }
+
+  async findUserById(user_id: string): Promise<object> {
+    const cacheFindUser: string = await this.cacheManager.get(
+      `userId_profile_${user_id}`,
+    );
+    if (cacheFindUser) {
+      const result = JSON.parse(cacheFindUser);
+      return result;
+    }
+    const userInfo = await this.userRepository.findOne({
+      where: { user_id },
+      relations: ['tags'],
+    });
+    if (!userInfo)
+      throw new HttpException('user id not found', HttpStatus.BAD_REQUEST);
+    console.log('資料庫找id info');
+    return userInfo;
+  }
+
+  async updateUserProfile(body: UserProfileDto, file: object): Promise<object> {
```

Comment:
Please add the unit test for this function

Reply:
will do it later